docs/en/resource/supplemental.md

```diff
@@ -0,0 +1,351 @@
+
+-   We should cater for filering via the query option.
+-   We can side load the full data set using the `put` param. This will allow data-aware widgets access to the full data set up front and is useful for widgets such as `select` which need to skip to content related to a given keypress etc...
+
+```ts
+import USStates from './states';
+
+const memoryTemplate: DataTemplate = {
+	read: ({ query }, put) => {
+		// filter the data
+		const filteredData = filter && query ? USStates.filter((i) => filter(query, i)) : USStates;
+		// put the filtered data into the resource so it
+		// won't try and read it again for the same query
+		put(0, filteredData);
+		// return the data that was requested
+		return { data: filteredData, total: filteredData.length };
+	}
+};
+```
+
+## Initialising a resource with local data
+
+When the local data you're using may change or is generated, an alternative approach is required to set up a memory template and pass data to the widget.
+
+```ts
+const memoryTemplate: DataTemplate = {
+	read: ({ query }, get, put) => {
+		// get the data from the resource
+		let data: any[] = get();
+
+		// filter the data
+		const filteredData = filter && query ? data.filter((i) => filter(query, i)) : data;
+		// put the filtered data into the resource so it
+		// won't try and read it again for the same query
+		put(0, filteredData);
+		// return the data that was requested
+		return { data: filteredData, total: filteredData.length };
+	}
+};
+
+export default factory(function() {
+	// pass a resource factory and the data to sideload
+	// instead of just the resource
+	return (
+		<List
+			resource={{
+				resource: () => createResource(memoryTemplate),
+				data: ['dog', 'fish', 'cat']
+			}}
+		/>
+	);
+});
+```
+
+The above approach will allow the `data` to be changed at any time causing the resource to be cleared and the new data sideloaded again. This is the approach that should be used in most in memory data array scenarios and can be seen being used within our `TimePicker` widget when time options are generated and passed to the `Select`.
+
+# Data Middleware
+
+The data middleware provides a widget with access to the underlying resource via the data middleware which offers an api to initialise and share the resource and functions to interact with it. The middleware should be imported and passed to the widgets `create` function the same as other middlwares. It will add two extra properties to your widget; `resource` and `transform`.
+
+```ts
+import { create } from '@dojo/framework/core/vdom';
+import { createDataMiddleware } from '@dojo/framework/core/middleware/data';
+
+const factory = create({ data });
+
+export const DataAwareWidget = factory(function ({
+	middleware: { data }
+}) {
+	const api = data();
+}
+```
+
+## The Resource Property
+
+The resource property added to a widget can be used to pass a resource, a resource factory and data to initialise the resource with or a resource wrapper.
+
+### resource
```

Comment:
Casing... it these are code then they can go in code guards otherwise they should be sentence case I believe.